core/src/main/java/com/linecorp/armeria/internal/common/RequestContextUtil.java

```diff
@@ -74,21 +138,69 @@ public static IllegalStateException newIllegalContextPushingException(
+     * the unexpected thread or forgetting to close the previous context.
+     */
+    public static IllegalStateException newIllegalContextPoppingException(
+            RequestContext currentCtx, RequestContext contextInStorage) {
+        requireNonNull(currentCtx, "currentCtx");
+        requireNonNull(contextInStorage, "contextInStorage");
+        final IllegalStateException ex = new IllegalStateException(
+                "The currentCtx " + currentCtx + " is not the same as the context in the storage: " +
+                contextInStorage + ". This means the callback was called from " +
+                "unexpected thread or forgetting to close previous context.");
+        if (REPORTED_THREADS.add(Thread.currentThread())) {
+            logger.warn("An error occurred while popping a context", ex);
+        }
+        return ex;
+    }
+
+    /**
+     * Returns the current {@link RequestContext} in the {@link RequestContextStorage}.
+     */
+    @Nullable
+    @SuppressWarnings("unchecked")
+    public static <T extends RequestContext> T get() {
+        return (T) requestContextStorage.currentOrNull();
+    }
+
+    /**
+     * Sets the specified {@link RequestContext} in the {@link RequestContextStorage} and
+     * returns the old {@link RequestContext}.
+     */
+    @Nullable
+    @SuppressWarnings("unchecked")
+    public static <T extends RequestContext> T getAndSet(RequestContext ctx) {
+        requireNonNull(ctx, "ctx");
+        return (T) requestContextStorage.push(ctx);
+    }
+
+    /**
+     * Removes the {@link RequestContext} in the {@link RequestContextStorage} if exists and returns
+     * {@link SafeCloseable} which pushes the {@link RequestContext} back to the {@link RequestContextStorage}.
      *
      * <p>Because this method pops the {@link RequestContext} arbitrarily, it shouldn't be used in
      * most cases. One of the examples this can be used is in {@link ChannelFutureListener}.
      * The {@link ChannelFuture} can be complete when the eventloop handles the different request. The
-     * eventloop might have the wrong {@link RequestContext} in the thread-local, so we should pop it.
+     * eventloop might have the wrong {@link RequestContext} in the {@link RequestContextStorage},
+     * so we should pop it.
      */
     public static SafeCloseable pop() {
-        final RequestContext oldCtx = RequestContextThreadLocal.getAndRemove();
+        final RequestContext oldCtx = requestContextStorage.currentOrNull();
         if (oldCtx == null) {
             return noopSafeCloseable();
         }
 
-        return () -> RequestContextThreadLocal.set(oldCtx);
+        pop(oldCtx, null);
+        return () -> requestContextStorage.push(oldCtx);
+    }
+
+    /**
+     * Pops the current {@link RequestContext} in the storage and pushes back the specified {@code toRestore}.
+     */
+    public static void pop(RequestContext current, @Nullable RequestContext toRestore) {
```

Comment:
How about renaming this `pop` to `replace`? Because It is hart to predict the behavior when I saw `pop` method with arguments.
```java
requestContextStorage.pop(current, toRestore)
```
However when I think this method name is `replace`, then I can understand the effect of the API.
```java
requestContextStorage.replace(current, toRestore)
```

Reply:
As we discussed, let's keep it as it is.